bin/liquidator/src/tx_sender.rs

```diff
@@ -0,0 +1,130 @@
+                let rebalance_trigger_sender = rebalance_trigger_sender.clone();
+                let mut liquidation = liquidation.clone();
+                let mut tcs = tcs.clone();
+                let id = i;
+                async move {
+                    loop {
+                        debug!("Worker #{} waiting for task", id);
+                        let _ = receiver.recv().await.unwrap();
+
+                        // a task is supposed to be available to process
+                        // find it in global shared state, and mark it as processing
+                        // (it's also possible we don't find anything in case liq/tcs job enqueued the same task multiple time)
+                        let (l, t) = {
+                            let mut writer = shared_state.write().unwrap();
+
+                            // print out list of all task for debugging
+                            for x in &writer.liquidation_candidates_accounts {
+                                if !writer.processing_liquidation.contains(x) {
+                                    tracing::trace!(" - LIQ {:?}", x);
+                                }
+                            }
+                            for x in &writer.interesting_tcs {
+                                if !writer.processing_tcs.contains(x) {
+                                    tracing::trace!("  - TCS {:?}", x);
+                                }
+                            }
+
+                            // next task to execute
+                            let liq_candidate = writer
+                                .liquidation_candidates_accounts
+                                .iter()
+                                .find(|x| !writer.processing_liquidation.contains(x))
+                                .map(|x| *x);
+                            let tcs_candidate = writer
+                                .interesting_tcs
+                                .iter()
+                                .find(|x| !writer.processing_tcs.contains(x))
+                                .map(|x| *x);
+
+                            if let Some(l) = liq_candidate {
+                                debug!("worker #{} got a liq candidate -> {}", id, l);
+                                writer.processing_liquidation.insert(l);
+                                (Some(l), None)
+                            } else if let Some(t) = tcs_candidate {
+                                debug!(
+                                    "worker #{} got a tcs candidate -> {:?} (out of {})",
+                                    id,
+                                    t,
+                                    writer.interesting_tcs.len()
+                                );
+                                writer.processing_tcs.insert(t);
+                                (None, Some(t))
+                            } else {
+                                debug!("worker #{} got nothing", id);
```

Comment:
this is indicative of a bug, isn't it? there's the assumption that each tx trigger channel message corresponds to one thing that needs to be done. If they don't match up it's possible that the channel is empty but there's still work to do.